x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/persistence/JobConfigProvider.java

```diff
@@ -0,0 +1,396 @@
+     * via the listener.
+     *
+     * @param jobId The job ID
+     * @param jobListener Job listener
+     */
+    public void getJob(String jobId, ActionListener<Job.Builder> jobListener) {
+        GetRequest getRequest = new GetRequest(AnomalyDetectorsIndex.configIndexName(),
+                ElasticsearchMappings.DOC_TYPE, Job.documentId(jobId));
+
+        executeAsyncWithOrigin(client, ML_ORIGIN, GetAction.INSTANCE, getRequest, new ActionListener<GetResponse>() {
+            @Override
+            public void onResponse(GetResponse getResponse) {
+                if (getResponse.isExists() == false) {
+                    jobListener.onFailure(ExceptionsHelper.missingJobException(jobId));
+                    return;
+                }
+
+                BytesReference source = getResponse.getSourceAsBytesRef();
+                parseJobLenientlyFromSource(source, jobListener);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                jobListener.onFailure(e);
+            }
+        });
+    }
+
+    /**
+     * Delete the anomaly detector job config document
+     *
+     * @param jobId The job id
+     * @param actionListener Deleted job listener
+     */
+    public void deleteJob(String jobId,  ActionListener<DeleteResponse> actionListener) {
+        DeleteRequest request = new DeleteRequest(AnomalyDetectorsIndex.configIndexName(),
+                ElasticsearchMappings.DOC_TYPE, Job.documentId(jobId));
+
+        executeAsyncWithOrigin(client, ML_ORIGIN, DeleteAction.INSTANCE, request, new ActionListener<DeleteResponse>() {
+            @Override
+            public void onResponse(DeleteResponse deleteResponse) {
+                if (deleteResponse.getResult() == DocWriteResponse.Result.NOT_FOUND) {
+                    actionListener.onFailure(ExceptionsHelper.missingJobException(jobId));
+                    return;
+                }
+
+                assert deleteResponse.getResult() == DocWriteResponse.Result.DELETED;
+                actionListener.onResponse(deleteResponse);
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                actionListener.onFailure(e);
+            }
+        });
+    }
+
+    /**
+     * Get the job and update it by applying {@code jobUpdater} then index the changed job
+     * setting the version in the request. Applying the update may cause a validation error
+     * which is returned via {@code updatedJobListener}
+     *
+     * @param jobId The Id of the job to update
+     * @param update The job update
+     * @param maxModelMemoryLimit The maximum model memory allowed
+     * @param updatedJobListener Updated job listener
+     */
+    public void updateJob(String jobId, JobUpdate update, ByteSizeValue maxModelMemoryLimit, ActionListener<Job> updatedJobListener) {
+        GetRequest getRequest = new GetRequest(AnomalyDetectorsIndex.configIndexName(),
```

Comment:
Could we reuse the `getJob` method here?

Reply:
No because the `version` is required to ensure consistency when the updated job is indexed.